File: sausage.cpp
// 2021 UCF HSPT - Sharon's Sausages
// Author: Billy Quiroga

#include <bits/stdc++.h>

using namespace std;

// max height that a sausage can be:
const int MAX = 100;

void solve() {
  int n;
  cin >> n;
  vector<int> s(n);
  for(int i = 0; i < n; i++)
    cin >> s[i];

  // If we are at index i, prefix_x[x] will hold the
  // number of times x appears before i.
  vector<long long> prefix_x(MAX + 1);

  // If we are at index i, suffix_x[x] will hold the
  // number of times x appears after i
  vector<long long> suffix_x(MAX + 1);

  // If we are at index i, prefix_xy[x][y] will hold the
  // number of times the pair (x, y) appears before i
  vector<vector<long long>> prefix_xy(MAX + 1, vector<long long>(MAX + 1));

  long long res = 0;

  // Build suffix_x since we are going to start at index 0
  // making the whole array the suffix
  for(int i = 0; i < n; i++)
    suffix_x[s[i]]++;

  // The idea for this solution is that since we are trying to
  // find subsequences of the form "x y y x", brute force the 
  // second y, while keeping track of the number of times
  // the subsequence "x y" appears before our current index
  // and the number of times "x" appears after our current index.
  // So with our current y, we know the number of ways to make this
  // the second y in a subsequence of the form "x y y x":
  // prefix_xy[x][y] * suffix_x[x]
  // It is this because we can choose any "x y" in the prefix and 
  // pair it up with any "x" in the suffix.
  for(int i = 0; i < n; i++) {

    // Let s[i] be the second y in our subsequence "x y y x"
    int y = s[i];

    // Remove s[i] from suffix_x because we only want
    // the frequency of s[i] after i, not including i
    suffix_x[s[i]]--;

    // Brute force the x we want to match our current y with
    // and calculate how many ways there are to do so.
    // Add that result to our final answer.
    for(int x = 1; x <= MAX; x++)
      res += prefix_xy[x][y] * suffix_x[x];

    // Update our current prefix_xy by brute forcing each x
    // we can pair our current y with. It should just be the
    // number of times x appears before our current y, which
    // is represented by prefix_x[x].
    for(int x = 1; x <= MAX; x++)
      prefix_xy[x][y] += prefix_x[x];

    // since s[i] will now be included in the prefix on the
    // next index, add it into prefix_x.
    prefix_x[s[i]]++;
  }

  cout << res << '\n';
}

int main() {
  int t;
  cin >> t;
  for(int i = 0; i < t; i++) solve();
  return 0;
}

File: sausage.py
# sausage.py by Jacob Steinebronn
# So we have an array and want to find the number of times
# "a b b a" is a subsequence, for any a and any b.
# The idea here is we brute force a, and keep track of every partial
# solution at some prefix (so how many as, abs, abbs, and abbas). 
# We assume that every index we look at is a b, and maybe an a if it matches.
# We have to keep partial solutions seperate since there are different bs, but
# we can agglomerate a, abb, and abba since those only depend on a (or nothing). 

numTests = (int)(input().strip())

for testNum in range(numTests):
    n = (int)(input().strip())
    arr = list(map(int, input().strip().split(" ")))

    ans = 0
    # Partial solutions of the form ab
    ab = [0]*105

    # Brute force a
    for a in range(1, 101, 1):
        # set up our partial solutions
        r_a = 0
        abb  = 0
        # clear the ab array
        for i in range(len(ab)):
            ab[i] = 0

        # For every element, advance partial solutions
        for x in arr:
            # If we have an a, we can complete every abb
            if x == a:
                ans += abb
            # For this b, we can make an abb from every ab
            abb += ab[x]
            # For this b, we can make an ab from every a
            ab[x] += r_a

            # If we have an a, we have yet another possible a
            if x == a:
                r_a += 1 

    print(ans)